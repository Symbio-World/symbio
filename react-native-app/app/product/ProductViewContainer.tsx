import * as React from 'react'
import {
  NO_SEARCH_RESULTS_FOUND,
  NO_USEFUL_INFO_FOUND,
} from '@symbio/barcode-processor-core'
import { isFailureOfType } from '@symbio/ts-lib'
import { ErrorViewContainer } from '../error-handler/ErrorViewContainer'
import { ProductView } from './ProductView'
import { ProductNotFound } from './ProductNotFound'
import { observeProductData } from './observeProductData'
import { ObservableView } from '../lib/ObservableView'

const isProductNotFound = (error: unknown) =>
  isFailureOfType(error, NO_SEARCH_RESULTS_FOUND) ||
  isFailureOfType(error, NO_USEFUL_INFO_FOUND)

type Props = {
  barcode: string
  onClose?: () => void
}
export const ProductViewContainer: React.FC<Props> = ({ barcode, onClose }) => {
  return (
    <ObservableView
      observable={observeProductData(barcode)}
      renderSuccess={(productData) => (
        <ProductView {...productData} onClose={onClose} />
      )}
      renderError={(error) =>
        isProductNotFound(error) ? (
          <ProductNotFound barcode={barcode} />
        ) : (
          <ErrorViewContainer error={error} onClose={onClose} />
        )
      }
    />
  )
}
