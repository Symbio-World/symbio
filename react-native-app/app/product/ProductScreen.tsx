import React from 'react'
import { RouteProp } from '@react-navigation/native'
import { StackNavigationProp } from '@react-navigation/stack'
import { ProductViewContainer } from './ProductViewContainer'
import { RootStackParamList } from '../Navigation'

type Props = {
  navigation: StackNavigationProp<RootStackParamList, 'ProductScreen'>
  route: RouteProp<RootStackParamList, 'ProductScreen'>
}
export const ProductScreen: React.FC<Props> = ({ route, navigation }) => {
  const { barcode } = route.params

  const handleFeedbackPress = (title: string) => {
    navigation.navigate('FeedbackScreen', { title })
  }

  return (
    <ProductViewContainer
      barcode={barcode}
      onFeedbackPress={handleFeedbackPress}
    />
  )
}
