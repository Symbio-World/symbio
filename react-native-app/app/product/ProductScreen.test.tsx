import * as React from 'react'
import { render } from 'react-native-testing-library'
import { ProductScreen } from './ProductScreen'
import { ProductViewContainer } from './ProductViewContainer'

describe('ProductScreen', () => {
  const barcode = 'barcode'
  const route: any = {
    params: {
      barcode,
    },
  }
  const navigation: any = {
    navigate: jest.fn(),
  }

  it('renders correctly', () => {
    const { toJSON } = render(
      <ProductScreen route={route} navigation={navigation} />,
    )
    expect(toJSON()).toMatchSnapshot()
  })

  it('passes barcode', () => {
    const { getByType } = render(
      <ProductScreen route={route} navigation={navigation} />,
    )
    expect(getByType(ProductViewContainer).props.barcode).toEqual(barcode)
  })
})
