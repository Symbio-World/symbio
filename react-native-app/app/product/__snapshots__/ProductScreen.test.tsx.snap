// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductScreen renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
      },
      Object {
        "justifyContent": "center",
      },
    ]
  }
>
  <Text>
    Something went wrong...
  </Text>
</View>
`;
